Tools/CISupport/ews-build/steps.py

```diff
@@ -7388,15 +7472,65 @@ def run(self):
-        if num_unexpected_results:
+        user_removed_tests = self.getProperty('user_removed_tests', [])
+        user_added_tests = self.getProperty('user_added_tests', [])
+
+        if user_removed_tests or user_added_tests:
+            unexpected_results_data = yield self.decode_results_data()
+            self.unexpected_results_filtered = unexpected_results_data
+            yield self.filter_results_by_user_modification(user_added_tests, user_removed_tests)
+
+        has_unexpected_results = self.getProperty('num_failing_files', 0) or self.getProperty('num_unexpected_issues', 0) or self.getProperty('num_passing_files', 0)
+        if has_unexpected_results and self.was_filtered:
+            yield self._addToLog('stdio', f'Found unexpected results after checking PR expectations!\n')
+            steps_to_add = [DownloadUnexpectedResultsFromMaster(), DeleteStaticAnalyzerResults(results_dir='StaticAnalyzerUnexpectedRegressions')]
+            steps_to_add += [GenerateSaferCPPResultsIndex(), DeleteStaticAnalyzerResults(), ArchiveStaticAnalyzerResults(), UploadStaticAnalyzerResults(), ExtractStaticAnalyzerTestResults(), DisplaySaferCPPResults()]
+            self.build.addStepsAfterCurrentStep(steps_to_add)
+        elif has_unexpected_results:
             self.build.addStepsAfterCurrentStep([ArchiveStaticAnalyzerResults(), UploadStaticAnalyzerResults(), ExtractStaticAnalyzerTestResults(), DisplaySaferCPPResults()])
 
         self.result_message = self.createResultMessage()
         return defer.returnValue(rc)
 
+    @defer.inlineCallbacks
+    def filter_results_by_user_modification(self, user_added_tests, user_removed_tests):
+        self.was_filtered = True
+        first_run_passes = self.getProperty('test_passes', [])
+        first_run_failures = self.getProperty('test_failures', [])
+
+        for test_name in user_added_tests:
+            split_test = test_name.split('/')
+            project, file, checker = split_test[0], '/'.join(split_test[1:-1]), split_test[-1]
+            yield self._addToLog('stdio', f'Filtering results for {test_name}:\n')
+            if test_name in first_run_passes and test_name not in self.unexpected_failures_filtered['passes'][project][checker]:
+                self.unexpected_results_filtered['passes'][project][checker].append(file)
+                yield self._addToLog('stdio', f'   Added {test_name} to unexpected passes.\n')
+            try:
+                self.unexpected_results_filtered['failures'][project][checker].remove(file)
+                yield self._addToLog('stdio', f'   Removed {test_name} from unexpected failures.\n')
+            except ValueError:
+                yield self._addToLog('stdio', f'   Did not remove {test_name} from unexpected failures since it did not fail.\n')
+
+        for test_name in user_removed_tests:
+            split_test = test_name.split('/')
+            project, file, checker = split_test[0], '/'.join(split_test[1:-1]), split_test[-1]
```

Comment:
should make sure that split_test[0] and split_test[1] etc. won't result in exception.